compiler/rustc_interface/src/util.rs

```diff
@@ -380,12 +388,17 @@ impl CodegenBackend for DummyCodegenBackend {
     ) {
         // JUSTIFICATION: TyCtxt no longer available here
         #[allow(rustc::bad_opt_access)]
-        if sess.opts.crate_types.iter().any(|&crate_type| crate_type != CrateType::Rlib) {
+        if codegen_results
+            .crate_info
+            .crate_types
+            .iter()
+            .any(|&crate_type| crate_type != CrateType::Rlib)
+        {
             #[allow(rustc::untranslatable_diagnostic)]
             #[allow(rustc::diagnostic_outside_of_impl)]
             sess.dcx().fatal(format!(
                 "crate type {} not supported by the dummy codegen backend",
-                sess.opts.crate_types[0],
+                codegen_results.crate_info.crate_types[0],
             ));
         }
```

Comment:
(pre-existing) This doesn't seem correct. Specifically, `crate_types[0]` is not necessarily the unsupported one. Using `.find` would be better here I think.

Also the check seems to disagree with the `supported_crate_types` above, is this deliberate?

Reply:
> Also the check seems to disagree with the supported_crate_types above, is this deliberate?

Yes, for `--crate-type bin` the dummy codegen backend can't produce an executable file, but tools like miri should still see the crate as an executable with `fn main()` if it exits before reaching the link step.

> (pre-existing) This doesn't seem correct. Specifically, crate_types[0] is not necessarily the unsupported one. Using .find would be better here I think.

It wasn't correct before, but now it technically is given that the only crate types we now have here are lib or bin which are mutually exclusive. I will change it to a find though as that is clearer.

Reply:
nit: you can technically use `{crate_type}` in the format string to make it a tad nicer.